KotlinNullPointerException in DataStore.sync

Context from thread in #gv: 

ddurst  5 days ago
1) We have a crash that by far outweighs everything else: kotlin.KotlinNullPointerException in mozilla.lockbox.store.DataStore.sync

ddurst  5 days ago
(and we see the corresponding jump in crashes when we released 1.1.3, which included almost exclusively new telemetry dispatches)

ddurst  5 days ago
2) I'm confused by the notion (which might be wrong) that when lockwise initiates a sync request, we don't really want that to be async -- but... it is, isn't it? Which means we could dispatch a sync_timeout (heck, even an sync_end) and that sync will actually succeed.
...
grisha:biking:  5 days ago
@ddurst so if i'm reading that code right, the timeout is certainly not necessary. It's not clear to me at a glance if a TimeoutException thrown on the scheduler will somehow negatively interact with the running sync - your comments imply that no, sync will continue on. To start, i'd remove that timeout logic entirely. Internally, sync library has its own limits for network timeouts and such, and depending on which version of a-s you're on, will return you with a very detailed SyncReport.

grisha:biking:  5 days ago
When you initiate a sync, you initiate an operation that may take quite some time (in both computer and human terms). We model this in Fenix via observables; you subscribe using a SyncStatusObserver, and are told when sync is running, when it stopped, when it failed, etc.

grisha:biking:  5 days ago
Medium-term plan for lockwise's sync & accounts implementation is to adopt a-c's high-level account & sync APIs - see https://github.com/mozilla-mobile/android-components/tree/master/components/service/firefox-accounts#high-level-apis-recommended-for-most-applications

grisha:biking:  5 days ago
I think the only blocker for that work is missing dataprotect integration for account storage in that high level API.

Stacktrace: 

```
Huawei P20 lite (HWANE), Android 9
Report 1
io.reactivex.exceptions.OnErrorNotImplementedException: 
  at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept (Functions.java:704)
  at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept (Functions.java:701)
  at io.reactivex.internal.observers.LambdaObserver.onError (LambdaObserver.java:77)
  at io.reactivex.internal.observers.LambdaObserver.onNext (LambdaObserver.java:67)
  at com.jakewharton.rxrelay2.ReplayRelay$SizeBoundReplayBuffer.replay (ReplayRelay.java:637)
  at com.jakewharton.rxrelay2.ReplayRelay.accept (ReplayRelay.java:220)
  at mozilla.lockbox.store.DataStore$unlockInternal$4.invoke (DataStore.java:182)
  at mozilla.lockbox.store.DataStore$unlockInternal$4.invoke (DataStore.java:39)
  at mozilla.lockbox.store.DataStore$sam$io_reactivex_functions_Consumer$0.accept (DataStore.java:2)
  at io.reactivex.internal.observers.LambdaObserver.onNext (LambdaObserver.java:63)
  at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext (ObservableMap.java:62)
  at io.reactivex.internal.operators.observable.ObservableTake$TakeObserver.onNext (ObservableTake.java:64)
  at io.reactivex.internal.operators.observable.ObservableSkip$SkipObserver.onNext (ObservableSkip.java:56)
  at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onNext (ObservableDoOnEach.java:101)
  at io.reactivex.internal.operators.observable.ObservableSwitchMap$SwitchMapObserver.drain (ObservableSwitchMap.java:297)
  at io.reactivex.internal.operators.observable.ObservableSwitchMap$SwitchMapInnerObserver.onNext (ObservableSwitchMap.java:374)
  at com.jakewharton.rxrelay2.BehaviorRelay$BehaviorDisposable.test (BehaviorRelay.java:354)
  at com.jakewharton.rxrelay2.BehaviorRelay$BehaviorDisposable.emitNext (BehaviorRelay.java:348)
  at com.jakewharton.rxrelay2.BehaviorRelay.accept (BehaviorRelay.java:127)
  at mozilla.lockbox.store.DataStore$updateList$1.accept (DataStore.java:254)
  at mozilla.lockbox.store.DataStore$updateList$1.accept (DataStore.java:39)
  at io.reactivex.internal.observers.ConsumerSingleObserver.onSuccess (ConsumerSingleObserver.java:62)
  at io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess (SingleCreate.java:67)
  at kotlinx.coroutines.rx2.RxSingleCoroutine.onCompleted (RxSingleCoroutine.java:45)
  at kotlinx.coroutines.AbstractCoroutine.onCompletionInternal (AbstractCoroutine.java:102)
  at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState (JobSupport.java:274)
  at kotlinx.coroutines.JobSupport.tryMakeCompleting (JobSupport.java:784)
  at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core (JobSupport.java:764)
  at kotlinx.coroutines.AbstractCoroutine.resumeWith (AbstractCoroutine.java:111)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java:46)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.java:238)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.java:594)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely (CoroutineScheduler.java:60)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.java:742)
Caused by: kotlin.KotlinNullPointerException: 
  at mozilla.lockbox.store.DataStore.sync (DataStore.java:220)
  at mozilla.lockbox.store.DataStore.syncIfRequired (DataStore.java:209)
  at mozilla.lockbox.store.DataStore.access$syncIfRequired (DataStore.java:39)
  at mozilla.lockbox.store.DataStore$1.accept (DataStore.java:92)
  at mozilla.lockbox.store.DataStore$1.accept (DataStore.java:39)
  at io.reactivex.internal.observers.LambdaObserver.onNext (LambdaObserver.java:63)
  at com.jakewharton.rxrelay2.ReplayRelay$SizeBoundReplayBuffer.replay (ReplayRelay.java:637)
  at com.jakewharton.rxrelay2.ReplayRelay.accept (ReplayRelay.java:220)
  at mozilla.lockbox.store.DataStore$unlockInternal$4.invoke (DataStore.java:182)
  at mozilla.lockbox.store.DataStore$unlockInternal$4.invoke (DataStore.java:39)
  at mozilla.lockbox.store.DataStore$sam$io_reactivex_functions_Consumer$0.accept (DataStore.java:2)
  at io.reactivex.internal.observers.LambdaObserver.onNext (LambdaObserver.java:63)
  at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext (ObservableMap.java:62)
  at io.reactivex.internal.operators.observable.ObservableTake$TakeObserver.onNext (ObservableTake.java:64)
  at io.reactivex.internal.operators.observable.ObservableSkip$SkipObserver.onNext (ObservableSkip.java:56)
  at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onNext (ObservableDoOnEach.java:101)
  at io.reactivex.internal.operators.observable.ObservableSwitchMap$SwitchMapObserver.drain (ObservableSwitchMap.java:297)
  at io.reactivex.internal.operators.observable.ObservableSwitchMap$SwitchMapInnerObserver.onNext (ObservableSwitchMap.java:374)
  at com.jakewharton.rxrelay2.BehaviorRelay$BehaviorDisposable.test (BehaviorRelay.java:354)
  at com.jakewharton.rxrelay2.BehaviorRelay$BehaviorDisposable.emitNext (BehaviorRelay.java:348)
  at com.jakewharton.rxrelay2.BehaviorRelay.accept (BehaviorRelay.java:127)
  at mozilla.lockbox.store.DataStore$updateList$1.accept (DataStore.java:254)
  at mozilla.lockbox.store.DataStore$updateList$1.accept (DataStore.java:39)
  at io.reactivex.internal.observers.ConsumerSingleObserver.onSuccess (ConsumerSingleObserver.java:62)
  at io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess (SingleCreate.java:67)
  at kotlinx.coroutines.rx2.RxSingleCoroutine.onCompleted (RxSingleCoroutine.java:45)
  at kotlinx.coroutines.AbstractCoroutine.onCompletionInternal (AbstractCoroutine.java:102)
  at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState (JobSupport.java:274)
  at kotlinx.coroutines.JobSupport.tryMakeCompleting (JobSupport.java:784)
  at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core (JobSupport.java:764)
  at kotlinx.coroutines.AbstractCoroutine.resumeWith (AbstractCoroutine.java:111)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java:46)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.java:238)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.java:594)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely (CoroutineScheduler.java:60)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.java:742)
```

--- a/app/src/main/java/mozilla/lockbox/store/DataStore.kt
+++ b/app/src/main/java/mozilla/lockbox/store/DataStore.kt
@@ -4,6 +4,7 @@
 
 package mozilla.lockbox.store
 
+import androidx.annotation.VisibleForTesting
 import com.jakewharton.rxrelay2.BehaviorRelay
 import com.jakewharton.rxrelay2.ReplayRelay
 import io.reactivex.Observable
@@ -60,7 +61,8 @@ open class DataStore(
 
     internal val compositeDisposable = CompositeDisposable()
     private val stateSubject = ReplayRelay.createWithSize<State>(1)
-    private val syncStateSubject = BehaviorRelay.createDefault<SyncState>(SyncState.NotSyncing)
+    @VisibleForTesting
+    val syncStateSubject: BehaviorRelay<SyncState> = BehaviorRelay.createDefault(SyncState.NotSyncing)
     private val listSubject: BehaviorRelay<List<ServerPassword>> = BehaviorRelay.createDefault(emptyList())
     private val deletedItemSubject = ReplayRelay.create<Consumable<ServerPassword>>()
 
@@ -235,13 +237,21 @@ open class DataStore(
         }
     }
 
-    private fun sync() {
+    @VisibleForTesting(
+        otherwise = VisibleForTesting.PRIVATE
+    )
+    fun sync() {
         resetSupport(support)
+
+        val syncConfig = support.syncConfig ?: run {
+            log.error("syncConfig is null in sync. This is likely a bug.")
+            return
+        }
 
         // ideally, we don't sync unless we are connected to the network
         syncStateSubject.accept(SyncState.Syncing)
 
-        backend.sync(support.syncConfig!!)
+        backend.sync(syncConfig)
             .asSingle(coroutineContext)
             .map {
                 log.debug("Hashed UID: $it")

--- a/app/src/test/java/mozilla/lockbox/store/DataStoreTest.kt
+++ b/app/src/test/java/mozilla/lockbox/store/DataStoreTest.kt
@@ -57,6 +57,15 @@ class DataStoreTest : DisposingTest() {
     }
 
     @Test
+    fun `test issue 867 sync crash`() {
+        support.syncConfig = null
+        Assert.assertNull(support.syncConfig)
+        subject.sync()
+        Assert.assertNull(support.syncConfig)
+        Assert.assertEquals(subject.syncStateSubject.value, DataStore.SyncState.NotSyncing)
+    }
+
+    @Test
     fun testLockUnlock_shouldSync() {
         val stateIterator = this.subject.state.blockingIterable().iterator()
         val listIterator = this.subject.list.blockingIterable().iterator()

